### PR TITLE
Add to_slice serialization function

### LINKED ISSUE #35: Serialize into externally allocated buffer
It would be nice to have the option of passing in a byte slice to be used in serialization, rather than the internal buffer.

Usecase example:
```rust
pub fn as_json(&self, buf: &mut [u8]) -> Result<usize, ()> {
    let json = serde_json_core::to_vec::<consts::U1024, _>(&self).map_err(|_e| ())?;
    let len = json.len();
    buf[..len].copy_from_slice(&json);
    Ok(len)
}
```

Would be awesome to be able to:
```rust
pub fn as_json(&self, buf: &mut [u8]) -> Result<usize, ()> {
    let len = serde_json_core::to_slice(&self, &buf).map_err(|_e| ())?;
    Ok(len)
}
```

Where `to_slice` would just populate `buf[0..len]` with the serialized output.

### PR DESCRIPTION
**Pull request by MathiasKoch**
- Add a `to_slice` function when serializing, that allows serialization into an external buffer, eliminating the need for an additional copy in some cases. 
- Change `to_vec` to make use of above introduced `to_slice`
- Change `to_string` to make use of `to_vec` with an additional utf8 check

Fixes #35

### CODE REVIEW

**Comment by MathiasKoch** [comment]
@japaric

File: src/ser/mod.rs [hunk 1]
```diff
@@ -51,19 +51,40 @@ impl fmt::Display for Error {
...
 }
 
-pub(crate) struct Serializer<B>
-where
-    B: heapless::ArrayLength<u8>,
-{
-    buf: Vec<u8, B>,
+pub(crate) struct Serializer<'a> {
+    buf: &'a mut [u8],
+    idx: usize,
```
Comment (eldruin) [inline src/ser/mod.rs:56]: Can this be called `current_length` or something like that?

File: src/ser/mod.rs [hunk 2]
```diff
@@ -86,7 +107,7 @@ macro_rules! serialize_unsigned {
             }
         }
 
-        $self.buf.extend_from_slice(&buf[i..])?;
+        $self.extend_from_slice(&buf[i..])?;
         Ok(())
```
Comment (eldruin) [inline src/ser/mod.rs:111]:
```suggestion
        $self.extend_from_slice(&buf[i..])
```

**Review by eldruin — COMMENTED** [review]
Other than the nitpick about the variable name, looks good to me, thanks!

**Comment by eldruin** [comment]
Also, could you add an entry to the changelog for this?

**Comment by eldruin** [comment]
Also, how does this differ from #27 ?

**Comment by MathiasKoch** [comment]
@eldruin I think you might be right. 

It looks like they add exactly the same functionality. So i think it is mostly a question of which approach do you like better?

**Comment by eldruin** [comment]
I think this version is simpler so let's merge this one and close #27

**Pull request merged**